heat/core/dndarray.py

```diff
@@ -1399,222 +1401,175 @@ def __getitem__(self, key):
         (1/2) >>> tensor([0.])
         (2/2) >>> tensor([0., 0.])
         """
-
+        # print('start getitem', key)
+        # print()
         l_dtype = self.dtype.torch_type()
-        if isinstance(key, DNDarray) and key.gshape[-1] != len(self.gshape):
-            key = tuple(x.item() for x in key)
+        kgshape_flag = False
+        if isinstance(key, DNDarray) and key.ndim == self.ndim:
+            lkey = [slice(None, None, None)] * self.ndim
+            kgshape_flag = True
+            kgshape = [0] * len(self.gshape)
+            if key.ndim > 1:
+                for i in range(key.ndim):
+                    kgshape[i] = key.gshape[i]
+                    lkey[i] = key._DNDarray__array[..., i]
+            else:
+                kgshape[0] = key.gshape[0]
+                lkey[0] = key._DNDarray__array
+            key = tuple(lkey)
+        elif not isinstance(key, tuple):
+            h = [slice(None, None, None)] * self.ndim
+            if isinstance(key, DNDarray):
+                h[0] = key._DNDarray__array.tolist()
+            elif isinstance(key, torch.Tensor):
+                h[0] = key.tolist()
+            else:
+                h[0] = key
+            key = tuple(h)
+
+        gout_full = [None] * len(self.gshape)
+        key = list(key)
+        for c, k in enumerate(key):
+            if isinstance(k, slice):
+                new_slice = stride_tricks.sanitize_slice(k, self.gshape[c])
+                gout_full[c] = math.ceil((new_slice.stop - new_slice.start) / new_slice.step)
+            elif isinstance(k, list):
+                gout_full[c] = len(k)
+            elif isinstance(k, (DNDarray, torch.Tensor)):
+                gout_full[c] = k.shape[0] if not kgshape_flag else kgshape[c]
+            if isinstance(k, DNDarray):
+                key[c] = k._DNDarray__array
+        if all([g == 1 for g in gout_full]):
```

Comment:
remove the square brackets, then the expression will become a generator and not copy to list afterwards